src/libraries/System.Text.Json/Common/JsonSourceGenerationOptionsAttribute.cs

```diff
@@ -40,6 +40,11 @@ public JsonSourceGenerationOptionsAttribute(JsonSerializerDefaults defaults)
             }
         }
 
+        /// <summary>
+        /// Specifies the default value of <see cref="JsonSerializerOptions.AllowOutOfOrderMetadataProperties"/> when set.
+        /// </summary>
+        public bool AllowOutOfOrderMetadataProperties { get; set; }
```

Comment:
> which would (correct me if I'm wrong) normally be an error

It might not be, for example it's possible to specify a type discriminator for a derived type but not the base type so its presence on the wire can be entirely optional.

Reply:
That is to say, we might decide in API review that we want this to be the default behavior (at the risk of overbuffering or even OOM'ing for scenaria that used to work in the past)

Reply:
> > which would (correct me if I'm wrong) normally be an error
> 
> It might not be, for example it's possible to specify a type discriminator for a derived type but not the base type so its presence on the wire can be entirely optional.

I still don't see how that affects the answer. If the necessary type discriminator is first, today it's valid, there's no buffering, and tomorrow there wouldn't be buffering either. If the necessary type discrimator is not first, today that's an error, and tomorrow there would be buffering but it would only be in cases where it was erroneous already; it wouldn't be making something that's currently fast instead be slow.

Where am I getting that wrong?

Reply:
> If the necessary type discriminator is first, today it's valid, there's no buffering, and tomorrow there wouldn't be buffering either. 

It would buffer even in that scenario since it would want to look ahead for other metadata properties such as `$id` or `$values`.

> Where am I getting that wrong?

Consider the hierarchy:

```C#
[JsonDerivedType(typeof(Derived), "derived"))]
public record Base(IEnumerable<int> values);
public record Derived(int x) : Base([x]);
```

This makes type discriminators legal because you have Derived but also optional since `Base` doesn't specify one. In that case you could today write something like
```C#
var value = new Base(Enumerable.Repeat(1, int.MaxValue));
using fs = File.OpenWrite("/some/path");
JsonSerializer.Serialize(fs, value); // { "values" : [1, 1, ... ] }
```
Currently the above can serialize and deserialize without any overbuffering whatsoever, however assuming we turned on reading ahead by default then it would try to buffer the entire payload so that it can look for the optional `$type` declaration.

Reply:
And that must be the behavior, that the last rather than first discriminator is used? If they're "allowed out of order", isn't that don't order doesn't matter and why can pick whichever?

Reply:
"out of order" in this context means making it possible to intersperse metadata properties among data. It doesn't have any impact on how duplicate properties are handled, in both modes an exception will be thrown eventually.

Per the example above, the problem comes from having to buffer when reading a valid encoding that doesn't have any metadata in its payload.

Reply:
It's by design. The presence of a type discriminator is optional (`JsonDerivedType` attribute annotations don't require it). This is to cater to polymorphic scenaria where emitting discriminator metadata is undesirable (and round tripping is not a concern). We could have required that annotations be consistent for a given hierarchy, but ultimately this can only be verified at run time and is less flexible (e.g. it would require all types in the open hierarchy to be annotated, including the base type itself). 

Ultimately though, there are more metadata properties than just `$type`. We'd still want to look ahead for potential `$values` for cases where some of the types in the hierarchy are collections, or `$id` for cases where we also have reference preservation enabled.